### PR TITLE
feat: add the state machine for sync states

### PR DESCRIPTION
**Pull request by afthabvp**
# Manage sync status via state machine

## Overview

This technical specification outlines the process for managing and tracking synchronization (sync) operations between various sources (e.g., BigQuery, PostgreSQL) and destinations (e.g., Slack, Facebook, Klaviyo). Sync operations play a crucial role in ensuring data is consistently and accurately transferred and updated across disparate platforms or services.

## Sync Status Definitions

Each sync run operation can be in one of the following states, which represent the sync run's current status:

- **Healthy**
  A state indicating the successful completion of a recent sync run operation without any issues.

- **Disabled**
  Indicates that the sync operation has been manually turned off and will not run until re-enabled.

- **Pending**
  Assigned immediately after a sync is set up, signaling that while the configuration is complete, no sync runs have been initiated.

- **Failed**
  Denotes a sync operation that encountered an error, preventing successful completion.

<Alert type="info">
  Note: Ensure that sync configurations are regularly reviewed to prevent prolonged periods in the Disabled or Failed states.
</Alert>

### Sync State Transitions

The following describes the allowed transitions between the sync states:

- **Pending ➔ Healthy**: Occurs when a sync run completes successfully.
- **Pending ➔ Failed**: Triggered if a sync run fails or is aborted.
- **Failed ➔ Healthy**: A successful sync run after a previously failed attempt.
- **Any state ➔ Disabled**: Reflects the manual disabling or enabling of the sync operation.

### Sync Run Status Definitions

- **Started**: Indicates that the sync operation has begun. This status serves as the initial state of a new sync run operation after being triggered.

- **Querying**: The sync is currently querying a source with its associated model to retrieve the latest data. This involves moving data to a temporary table called "SyncRecord".

- **Queued**: Indicates the sync is scheduled for execution, following the successful transfer of source data to the "SyncRecord" table. This marks the completion of the preparation phase, with the sync now ready to transmit data to the destination as per system scheduling and resource availability.

- **In Progress**: The sync is actively transferring data from the "SyncRecord" table to the destination. This phase marks the actual update or insertion of data into the destination database, reflecting the final step of the sync process.

- **Success**: The sync run is completed successfully without any issues.

- **Paused**: Indicates a temporary interruption occurred while transferring data from the "SyncRecord" table to the destination. The sync is paused but designed to automatically resume in a subsequent run, ensuring continuity of the sync process.

- **Aborted/Failed**: The sync has encountered an error that prevents it from completing successfully.

###  Sync Run State Transitions

The following describes the allowed transitions between the sync run states:

- **Started ➔ Querying**: Transition post-initiation as data retrieval begins.
- **Querying ➔ Queued**: After staging data in the "SyncRecord" table, indicating readiness for transmission.
- **Queued ➔ In Progress**: Commences as the sync operation begins writing data to the destination, based on availability of system resources.
- **In Progress ➔ Success**: Marks the successful completion of data transmission.
- **In Progress ➔ Paused**: Triggered by a temporary interruption in the sync process.
- **Paused ➔ In Progress**: Signifies the resumption of a sync operation post-interruption.
- **In Progress ➔ Aborted/Failed**: Initiated when an error prevents the successful completion of the sync operation.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would impact existing functionality)
- [ ] Documentation update
- [ ] Chore
  
## Checklist:
- [x] Ensure a branch name is prefixed with `feature`, `bugfix`, `hotfix`, `release` or `chore` followed by `/` and branch name e.g `feature/add-salesforce-connector`
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/models/sync_run.rb [hunk 1]
```diff
@@ -24,12 +25,50 @@ class SyncRun < ApplicationRecord
...
+    state :pending, initial: true
+    state :started
+    state :querying
+    state :queued
+    state :in_progress
+    state :success
+    state :paused
+    state :failed
+
+    event :start do
```
[the remark this answers is not in the document]
Comment (afthabvp) [reply app/models/sync_run.rb:38]: it should present tense

File: app/models/sync_run.rb [hunk 2]
```diff
@@ -24,12 +25,52 @@ class SyncRun < ApplicationRecord
...
+    event :queue do
+      transitions from: :querying, to: :queued
+    end
+
+    event :progress do
+      transitions from: %i[queued paused in_progress], to: :in_progress
+    end
+
+    event :pause do
+      transitions from: :in_progress, to: :paused
```
Comment (subintp) [inline app/models/sync_run.rb:57]: What stage pause will be called ?

File: app/temporal/activities/extractor_activity.rb [hunk 3]
```diff
@@ -10,22 +11,33 @@ class ExtractorActivity < Temporal::Activity
...
+      sync_run.start
+      sync_run.update(started_at: Time.zone.now)
 
       # TODO: Select extraction strategy
       # based on sync mode eg: incremental/full_refresh
       extractor = ReverseEtl::Extractors::IncrementalDelta.new
       extractor.read(sync_run.id, activity)
     end
+
+    def update_failure(sync_run)
```
Comment (subintp) [inline app/temporal/activities/extractor_activity.rb:37]: move this method to sync_run. Sync should marked as failure whenever we abort the sync run

File: app/temporal/activities/extractor_activity.rb [hunk 4]
```diff
@@ -10,22 +11,33 @@ class ExtractorActivity < Temporal::Activity
...
 
     def execute(sync_run_id)
       sync_run = SyncRun.find(sync_run_id)
 
-      sync_run.status = "in_progress"
-      sync_run.started_at = Time.zone.now
-      sync_run.save!
+      unless sync_run.may_start?
+        update_failure(sync_run)
+        raise SyncRunStateExeption,
```
Comment (subintp) [inline app/temporal/activities/extractor_activity.rb:23]: We need to confirm that the loader and reporter activity will be executed after raising this error. Better to write a test case for this.

Comment (subintp) [inline app/temporal/activities/extractor_activity.rb:23]: Skip the workflow if its is already executed.

File: app/temporal/activities/reporter_activity.rb [hunk 5]
```diff
@@ -27,10 +32,16 @@ def fetch_record_counts(sync_run)
       [total, success, failed]
     end
 
-    def determine_status(failed_rows, total_rows)
-      # TODO: Update status as incomplete if sync run retry exhausted
-      # Sync failure should be marked based on threshold failure percentage
-      failed_rows == total_rows ? :failed : :success
+    def update_sucess(sync_run)
```
Comment (subintp) [inline app/temporal/activities/reporter_activity.rb:35]: Move both methods to sync_run. Sync run is best place to own this logic

File: app/temporal/activities/reporter_activity.rb [hunk 6]
```diff
@@ -27,10 +32,16 @@ def fetch_record_counts(sync_run)
...
+      sync_run.complete!
+      sync = sync_run.sync
+      sync.complete!
+    end
+
+    def update_failure(sync_run)
```
Comment (subintp) [inline app/temporal/activities/reporter_activity.rb:41]: Same as above

File: app/temporal/activities/reporter_activity.rb [hunk 7]
```diff
@@ -5,16 +5,21 @@ class ReporterActivity < Temporal::Activity
     def execute(sync_run_id)
       sync_run = SyncRun.find(sync_run_id)
 
-      return unless sync_run.in_progress?
+      unless sync_run.may_complete?
+        update_failure(sync_run)
+        raise StandardError,
```
Comment (subintp) [inline app/temporal/activities/reporter_activity.rb:10]: This should non-retryable error. Nothing can be done if we reach this scenario.

File: lib/reverse_etl/extractors/incremental_delta.rb [hunk 8]
```diff
@@ -7,17 +7,26 @@ class IncrementalDelta < Base
 
       # TODO: Make it as class method
       def read(sync_run_id, activity)
-        sync_run = setup_sync_run(sync_run_id)
+        total_query_count = 0
+        sync_run = SyncRun.find(sync_run_id)
+
+        return unless sync_run.may_query?
```
Comment (subintp) [inline lib/reverse_etl/extractors/incremental_delta.rb:13]: Log this error and return

File: lib/reverse_etl/loaders/standard.rb [hunk 9]
```diff
@@ -6,6 +6,16 @@ class Standard < Base
       THREAD_COUNT = (ENV["SYNC_LOADER_THREAD_POOL_SIZE"] || "5").to_i
       def write(sync_run_id, activity)
         sync_run = SyncRun.find(sync_run_id)
+
+        unless sync_run.may_progress?
```
Comment (subintp) [inline lib/reverse_etl/loaders/standard.rb:10]: Skip the activity. This scenario can happen when a workflow is retried after an error in reporter activity

File: lib/reverse_etl/loaders/standard.rb [hunk 10]
```diff
@@ -83,6 +93,12 @@ def heartbeat(activity)
         activity.heartbeat
         raise StandardError, "Cancel activity request received" if activity.cancel_requested
       end
+
+      def update_failure(sync_run)
```
Comment (subintp) [inline lib/reverse_etl/loaders/standard.rb:97]: Move this method to sync_run

**Review by subintp — CHANGES_REQUESTED** [review]

File: app/temporal/activities/extractor_activity.rb (hunk 4 above)
Comment (afthabvp) [reply app/temporal/activities/extractor_activity.rb]: No need exception, we are logging error and skipping

File: app/models/sync_run.rb [hunk 11]
```diff
@@ -24,12 +25,52 @@ class SyncRun < ApplicationRecord
...
+    state :querying
+    state :queued
+    state :in_progress
+    state :success
+    state :paused
+    state :failed
+
+    event :start do
+      transitions from: :pending, to: :started
+      transitions from: :started, to: :started
```
[the remark this answers is not in the document]
Comment (afthabvp) [reply app/models/sync_run.rb]: pending => started , actual scenario
started => started,  early failure before moving to querying  and retry
querying => started,  when a retry with error in source query

File: app/controllers/api/v1/sync_runs_controller.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class SyncRunsController < ApplicationController
+      skip_before_action :validate_contract
+      def index
+        sync_runs = current_workspace.syncs.find(params[:sync_id]).sync_runs
```
[the remark this answers is not in the document]
Comment (afthabvp) [reply app/controllers/api/v1/sync_runs_controller.rb]: different pr

File: app/controllers/api/v1/sync_records_controller.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class SyncRecordsController < ApplicationController
+      skip_before_action :validate_contract
+      def index
+        sync_records = current_workspace.syncs.find(params[:sync_id]).sync_runs.find(params[:sync_run_id]).sync_records
```
[the remark this answers is not in the document]
Comment (afthabvp) [reply app/controllers/api/v1/sync_records_controller.rb]: different pr

File: app/models/sync.rb [hunk 14]
```diff
@@ -45,6 +47,29 @@ class Sync < ApplicationRecord
 
   default_scope { order(updated_at: :desc) }
 
+  aasm column: "status" do
+    state :pending, initial: true
+    state :healthy
+    state :failed
+    state :disabled
+
+    event :complete_successfully do
```
[the remark this answers is not in the document]
Comment (afthabvp) [reply app/models/sync.rb]: it should present tense

File: app/models/sync.rb [hunk 15]
```diff
@@ -45,6 +47,29 @@ class Sync < ApplicationRecord
...
+    state :healthy
+    state :failed
+    state :disabled
+
+    event :complete do
+      transitions from: :pending, to: :healthy
+    end
+
+    event :fail do
+      transitions from: :pending, to: :failed
```
[the remark this answers is not in the document]
Comment (afthabvp) [reply app/models/sync.rb]: done

File: app/temporal/activities/extractor_activity.rb (hunk 3 above)
Comment (afthabvp) [reply app/temporal/activities/extractor_activity.rb]: done

File: lib/reverse_etl/loaders/standard.rb (hunk 10 above)
Comment (afthabvp) [reply lib/reverse_etl/loaders/standard.rb]: done

File: app/models/sync_run.rb [hunk 16]
```diff
@@ -24,12 +25,55 @@ class SyncRun < ApplicationRecord
...
+    state :started
+    state :querying
+    state :queued
+    state :in_progress
+    state :success
+    state :paused
+    state :failed
+
+    event :start do
+      transitions from: %i[pending started querying], to: :started
```
Comment (karthik-sivadas) [inline app/models/sync_run.rb:39]: Why do we need started state in from transition? And what happens if it's queued? Can't a sync run move from queued state to started? Please write tests for these cases.

File: app/temporal/activities/extractor_activity.rb [hunk 17]
```diff
@@ -11,21 +11,30 @@ class ExtractorActivity < Temporal::Activity
...
 
     def execute(sync_run_id)
       sync_run = SyncRun.find(sync_run_id)
 
-      sync_run.status = "in_progress"
-      sync_run.started_at = Time.zone.now
-      sync_run.save!
+      return log_error_and_return(sync_run) unless sync_run.may_start?
+
+      # state of sync run to started only if current stete in [ pending,started,querying]
```
Comment (subintp) [inline app/temporal/activities/extractor_activity.rb:22]: typo in state

File: app/temporal/activities/reporter_activity.rb [hunk 18]
```diff
@@ -27,10 +27,12 @@ def fetch_record_counts(sync_run)
       [total, success, failed]
     end
 
-    def determine_status(failed_rows, total_rows)
-      # TODO: Update status as incomplete if sync run retry exhausted
-      # Sync failure should be marked based on threshold failure percentage
-      failed_rows == total_rows ? :failed : :success
+    def log_error_and_return(sync_run)
```
Comment (subintp) [inline app/temporal/activities/reporter_activity.rb:30]: why return in the function name ? This method only logs error

File: lib/reverse_etl/extractors/incremental_delta.rb [hunk 19]
```diff
@@ -7,17 +7,26 @@ class IncrementalDelta < Base
 
       # TODO: Make it as class method
       def read(sync_run_id, activity)
-        sync_run = setup_sync_run(sync_run_id)
+        total_query_count = 0
+        sync_run = SyncRun.find(sync_run_id)
+
+        return log_error_and_return(sync_run) unless sync_run.may_query?
```
Comment (subintp) [inline lib/reverse_etl/extractors/incremental_delta.rb:13]: return is not needed in the function name ?

File: lib/reverse_etl/extractors/incremental_delta.rb [hunk 20]
```diff
@@ -85,6 +87,14 @@ def update_or_create_sync_record(sync_record, record, sync_run, fingerprint)
         )
         sync_record.save!
       end
+
+      def log_error_and_return(sync_run)
```
Comment (subintp) [inline lib/reverse_etl/extractors/incremental_delta.rb:91]: why return in the method name ?

File: lib/reverse_etl/loaders/standard.rb [hunk 21]
```diff
@@ -6,6 +6,12 @@ class Standard < Base
       THREAD_COUNT = (ENV["SYNC_LOADER_THREAD_POOL_SIZE"] || "5").to_i
       def write(sync_run_id, activity)
         sync_run = SyncRun.find(sync_run_id)
+
+        return log_error_and_return(sync_run) unless sync_run.may_progress?
```
Comment (subintp) [inline lib/reverse_etl/loaders/standard.rb:10]: log_error ?

File: app/models/sync_run.rb [hunk 22]
```diff
@@ -24,12 +25,55 @@ class SyncRun < ApplicationRecord
...
+    event :query do
+      transitions from: :started, to: :querying
+    end
+
+    event :queue do
+      transitions from: :querying, to: :queued
+    end
+
+    event :progress do
+      transitions from: %i[queued paused in_progress], to: :in_progress
```
Comment (karthik-sivadas) [inline app/models/sync_run.rb:51]: Why are we going from inprogress to inprogress state

File: app/models/sync_run.rb [hunk 23]
```diff
@@ -24,12 +25,55 @@ class SyncRun < ApplicationRecord
...
+    event :pause do
+      transitions from: :in_progress, to: :paused
+    end
+
+    event :complete do
+      transitions from: :in_progress, to: :success
+    end
+
+    event :abort do
+      transitions from: %i[pending started querying queued in_progress paused failed], to: :failed
```
Comment (karthik-sivadas) [inline app/models/sync_run.rb:63]: Same here why are we adding a transitioning state here?

File: spec/lib/reverse_etl/loaders/standard_spec.rb [hunk 24]
```diff
@@ -128,5 +145,26 @@
         subject.write(sync_run_individual.id, activity)
       end
     end
+
+    context "when sync run not start" do
```
Comment (subintp) [inline spec/lib/reverse_etl/loaders/standard_spec.rb:149]: Suggestion: "Skip loading when status is corrupted"

File: spec/lib/reverse_etl/loaders/standard_spec.rb [hunk 25]
```diff
@@ -128,5 +145,26 @@
...
+      tracker = Multiwoven::Integrations::Protocol::TrackingMessage.new(
+        success: 0,
+        failed: 1
+      )
+      let(:transformer) { ReverseEtl::Transformers::UserMapping.new }
+      let(:transform) { transformer.transform(sync_individual, sync_record_individual) }
+      let(:multiwoven_message) { tracker.to_multiwoven_message }
+      let(:client) { instance_double(sync_individual.destination.connector_client) }
+
+      it "trying to started to in_progrss" do
```
Comment (subintp) [inline spec/lib/reverse_etl/loaders/standard_spec.rb:159]: Test message is not clear. Typo in in_progress

File: spec/temporal/activities/extractor_activity_spec.rb [hunk 26]
```diff
@@ -4,10 +4,74 @@
...
+      expect(sync_run.workspace_id).to eq(sync.workspace_id)
+      expect(sync_run.source_id).to eq(sync.source_id)
+      expect(sync_run.destination_id).to eq(sync.destination_id)
+      expect(sync_run.model_id).to eq(sync.model_id)
+    end
+
+    it "sync run started to started" do
+      expect(sync_run_satrted).to have_state(:started)
+      activity.execute(sync_run_satrted.id)
+      sync_run_satrted.reload
```
Comment (subintp) [inline spec/temporal/activities/extractor_activity_spec.rb:47]: typo

File: spec/temporal/activities/extractor_activity_spec.rb [hunk 27]
```diff
@@ -4,10 +4,74 @@
...
   describe "#execute" do
-    subject { described_class.new }
+    let(:source) do
+      create(:connector, connector_type: "source", connector_name: "Snowflake")
+    end
+    let(:destination) { create(:connector, connector_type: "destination") }
+    let!(:catalog) { create(:catalog, connector: destination) }
+    let!(:sync) { create(:sync, sync_interval: 3, sync_interval_unit: "hours", source:, destination:) }
+    let(:sync_run) { create(:sync_run, sync:, workspace: sync.workspace, source:, destination:, model: sync.model) }
+    let(:sync_run_satrted) do
```
Comment (subintp) [inline spec/temporal/activities/extractor_activity_spec.rb:14]: typo in variable naming

**Review by subintp — CHANGES_REQUESTED** [review]

File: spec/temporal/activities/extractor_activity_spec.rb [hunk 28]
```diff
@@ -4,10 +4,74 @@
...
+      expect(sync_run_querying.sync_id).to eq(sync.id)
+      expect(sync_run_querying.workspace_id).to eq(sync.workspace_id)
+      expect(sync_run_querying.source_id).to eq(sync.source_id)
+      expect(sync_run_querying.destination_id).to eq(sync.destination_id)
+      expect(sync_run_querying.model_id).to eq(sync.model_id)
+    end
 
-    it "executes without errors" do
-      # TODO: Add more tests
+    context "when sync run not start" do
```
Comment (subintp) [inline spec/temporal/activities/extractor_activity_spec.rb:68]: context message is not clear. "Skip extraction"

**Review by subintp — CHANGES_REQUESTED** [review]

File: app/temporal/activities/extractor_activity.rb [hunk 29]
```diff
@@ -11,21 +11,30 @@ class ExtractorActivity < Temporal::Activity
...
+      sync_run.start
+      sync_run.update(started_at: Time.zone.now)
 
       # TODO: Select extraction strategy
       # based on sync mode eg: incremental/full_refresh
       extractor = ReverseEtl::Extractors::IncrementalDelta.new
       extractor.read(sync_run.id, activity)
     end
+
+    def log_error_and_return(sync_run)
```
Comment (karthik-sivadas) [inline app/temporal/activities/extractor_activity.rb:32]: Shouldn't we change the state if it fails?

File: app/temporal/activities/reporter_activity.rb (hunk 18 above)
Comment (karthik-sivadas) [inline app/temporal/activities/reporter_activity.rb:30]: Same here we need to add state change for failure.

File: app/models/sync_run.rb (hunk 16 above)
Comment (afthabvp) [reply app/models/sync_run.rb:39]: If a temporal activity fails while in the "Started" state, a temporal retry may be initiated. For this retry to occur, the activity transitions from the "Started" state back to "Started," indicating a reattempt of the operation.

File: app/models/sync_run.rb (hunk 22 above)
Comment (afthabvp) [reply app/models/sync_run.rb:51]: same here, we need a transition from in_progress => in_progress for temporal activity retry

File: app/models/sync_run.rb (hunk 23 above)
Comment (afthabvp) [reply app/models/sync_run.rb:63]: same here, we need a transition from failed => failed for temporal activity retry

File: app/temporal/activities/extractor_activity.rb (hunk 29 above)
Comment (afthabvp) [reply app/temporal/activities/extractor_activity.rb:32]: If a sync run fails while in the "Queued" state, with all records present in "SyncRecord," and a retry is initiated from the beginning, Temporal needs to bypass all activities leading up to the "Queued" state during the retry process.

File: spec/temporal/activities/extractor_activity_spec.rb (hunk 27 above)
Comment (afthabvp) [reply spec/temporal/activities/extractor_activity_spec.rb]: done

File: spec/lib/reverse_etl/loaders/standard_spec.rb (hunk 25 above)
Comment (afthabvp) [reply spec/lib/reverse_etl/loaders/standard_spec.rb]: done

File: spec/lib/reverse_etl/loaders/standard_spec.rb (hunk 24 above)
Comment (afthabvp) [reply spec/lib/reverse_etl/loaders/standard_spec.rb]: done

File: app/temporal/activities/extractor_activity.rb (hunk 17 above)
Comment (afthabvp) [reply app/temporal/activities/extractor_activity.rb]: done

File: app/temporal/activities/reporter_activity.rb (hunk 18 above)
Comment (afthabvp) [reply app/temporal/activities/reporter_activity.rb]: done

File: lib/reverse_etl/extractors/incremental_delta.rb (hunk 20 above)
Comment (afthabvp) [reply lib/reverse_etl/extractors/incremental_delta.rb]: done

File: lib/reverse_etl/extractors/incremental_delta.rb (hunk 19 above)
Comment (afthabvp) [reply lib/reverse_etl/extractors/incremental_delta.rb]: done

File: lib/reverse_etl/loaders/standard.rb (hunk 21 above)
Comment (afthabvp) [reply lib/reverse_etl/loaders/standard.rb]: done

**Review by subintp — APPROVED** [review]

**Pull request merged**